### ISSUE TITLE
DNS to 1.1.1.1 is required for tunnel start

### ISSUE THREAD

**Issue by tmsteen** [cloudflare/cloudflared#269]
Cloudflared seems to require DNS to 1.1.1.1 instead of using system DNS settings. For restrictive networks, this is an additional egress consideration that is not listed in the documentation. 

Is there a way to force cloudflared to use system DNS settings or is this something that is on the roadmap?

**Comment by adamchalmers** [comment]
Seems like a good feature worth supporting. We'll add it to the backlog.

**Comment by adamchalmers** [comment]
Hi Trevor,

cloudflared shouldn't require 1.1.1.1, if you take a look at the [code where we resolve the Cloudflare edge](https://github.com/cloudflare/cloudflared/blob/master/edgediscovery/allregions/discovery.go#L62-L64) you'll see we try the system resolver, and if that fails, we fall back to 1.1.1.1.

We could make this choice of fallback DNS provider configurable, but I think most users should be fine with just configuring their system DNS, which cloudflared uses. Could you please check whether your system DNS can resolve the Cloudflare edge by running `dig region1.argotunnel.com`?

**Comment by tmsteen** [comment]
Thanks for the update. It does indeed resolve with the system resolver. Does cloudflared need to resolve the Argo Tunnel hostname? If so, I wonder if the issue we are having is this...

1. We provision a device with Argo and then send it to a client. 
2. Depending on how long it takes the client to plug the device in, the tunnel DNS record may be removed due to inactivity.
3. Client plugs in device, DNS record is created when Argo tried to reconnect but takes time to propagate, so resolution using the system resolver fails and falls back to 1.1.1.1.

The case that prompted me to open this issue was the the tunnel would not connect until we opened the firewall to allow 1.1.1.1:53 based on block logs from our client.

**Comment by adamchalmers** [comment]
Interesting. Could you paste some logs from a machine where 1.1.1.1 is still blocked? I'd like to see which step fails. And I'd also like to know how you're invoking cloudflared (any config file or CLI args or cloudflared-specific env vars) so I can figure out what DNS queries are being made.

**Comment by tmsteen** [comment]
```
/var/log/syslog.7.gz:Dec  2 10:03:04 ipentestdevice systemd[1]: cloudflared.service: Service hold-off time over, scheduling restart.
/var/log/syslog.7.gz:Dec  2 10:03:04 ipentestdevice systemd[1]: cloudflared.service: Scheduled restart job, restart counter is at 52.
/var/log/syslog.7.gz:Dec  2 10:03:04 ipentestdevice cloudflared[6082]: #033[36mINFO#033[0m[2020-12-02T10:03:04-06:00] Version 2020.11.11
/var/log/syslog.7.gz:Dec  2 10:03:04 ipentestdevice cloudflared[6082]: #033[36mINFO#033[0m[2020-12-02T10:03:04-06:00] GOOS: linux, GOVersion: go1.15.5, GoArch: amd64
/var/log/syslog.7.gz:Dec  2 10:03:04 ipentestdevice cloudflared[6082]: #033[36mINFO#033[0m[2020-12-02T10:03:04-06:00] Environment variables map[config:/etc/cloudflared/config.yml hostname:<redacted> no-autoupdate:true origincert:/etc/cloudflared/cert.pem proxy-dns-upstream:https://1.1.1.1/dns-query, https://1.0.0.1/dns-query url:ssh://127.0.0.1:22]
/var/log/syslog.7.gz:Dec  2 10:03:04 ipentestdevice cloudflared[6082]: #033[36mINFO#033[0m[2020-12-02T10:03:04-06:00] Starting metrics server on 127.0.0.1:46103/metrics
/var/log/syslog.7.gz:Dec  2 10:03:04 ipentestdevice cloudflared[6082]: #033[36mINFO#033[0m[2020-12-02T10:03:04-06:00] Initial protocol h2mux
/var/log/syslog.7.gz:Dec  2 10:03:04 ipentestdevice cloudflared[6082]: #033[31mERROR#033[0m[2020-12-02T10:03:04-06:00] Error looking up Cloudflare edge IPs: the DNS query failed: lookup argotunnel.com on 127.0.0.53:53:
 cannot unmarshal DNS message
/var/log/syslog.7.gz:Dec  2 10:03:04 ipentestdevice cloudflared[6082]: #033[31mERROR#033[0m[2020-12-02T10:03:04-06:00] Please try the following things to diagnose this issue:
/var/log/syslog.7.gz:Dec  2 10:03:04 ipentestdevice cloudflared[6082]: #033[31mERROR#033[0m[2020-12-02T10:03:04-06:00]   1. ensure that argotunnel.com is returning "origintunneld" service records.
/var/log/syslog.7.gz:Dec  2 10:03:04 ipentestdevice cloudflared[6082]: #033[31mERROR#033[0m[2020-12-02T10:03:04-06:00]      Run your system's equivalent of: dig srv _origintunneld._tcp.argotunnel.com
/var/log/syslog.7.gz:Dec  2 10:03:04 ipentestdevice cloudflared[6082]: #033[31mERROR#033[0m[2020-12-02T10:03:04-06:00]   2. ensure that your DNS resolver is not returning compressed SRV records.
/var/log/syslog.7.gz:Dec  2 10:03:04 ipentestdevice cloudflared[6082]: #033[31mERROR#033[0m[2020-12-02T10:03:04-06:00]      See GitHub issue https://github.com/golang/go/issues/27546
/var/log/syslog.7.gz:Dec  2 10:03:04 ipentestdevice cloudflared[6082]: #033[31mERROR#033[0m[2020-12-02T10:03:04-06:00]      For example, you could use Cloudflare's 1.1.1.1 as your resolver:
/var/log/syslog.7.gz:Dec  2 10:03:04 ipentestdevice cloudflared[6082]: #033[31mERROR#033[0m[2020-12-02T10:03:04-06:00]      https://developers.cloudflare.com/1.1.1.1/setting-up-1.1.1.1/
/var/log/syslog.7.gz:Dec  2 10:03:04 ipentestdevice cloudflared[6082]: #033[36mINFO#033[0m[2020-12-02T10:03:04-06:00] Initiating graceful shutdown due to Could not lookup srv records on _origintunneld._tcp.argotunnel.c
om: lookup argotunnel.com on 127.0.0.53:53: cannot unmarshal DNS message ...
/var/log/syslog.7.gz:Dec  2 10:03:04 ipentestdevice cloudflared[6082]: #033[31mERROR#033[0m[2020-12-02T10:03:04-06:00] Quitting due to error: Could not lookup srv records on _origintunneld._tcp.argotunnel.com: lookup a
rgotunnel.com on 127.0.0.53:53: cannot unmarshal DNS message
/var/log/syslog.7.gz:Dec  2 10:03:05 ipentestdevice cloudflared[6082]: #033[36mINFO#033[0m[2020-12-02T10:03:05-06:00] Metrics server stopped
/var/log/syslog.7.gz:Dec  2 10:03:05 ipentestdevice cloudflared[6082]: Could not lookup srv records on _origintunneld._tcp.argotunnel.com: lookup argotunnel.com on 127.0.0.53:53: cannot unmarshal DNS message
/var/log/syslog.7.gz:Dec  2 10:03:05 ipentestdevice systemd[1]: cloudflared.service: Main process exited, code=exited, status=1/FAILURE
/var/log/syslog.7.gz:Dec  2 10:03:05 ipentestdevice systemd[1]: cloudflared.service: Failed with result 'exit-code'.
```

This repeats every 5 seconds until we finally made the firewall change. I did notice this in there `proxy-dns-upstream:https://1.1.1.1/dns-query, https://1.0.0.1/dns-query` though we do not set that anywhere in our config file and I honestly don't know what effect that has. Config file contents:

```
hostname: <redacted>
url: ssh://127.0.0.1:22
```

We are also still using the `legacy` flag at the moment.

**Comment by adamchalmers** [comment]
So, if you take a look at [the code that produces that error](https://github.com/cloudflare/cloudflared/blob/master/edgediscovery/allregions/discovery.go#L62-L65) it seems the Go stdlib functions are having trouble using your system DNS to resolve the SRV (_origintunneld._tcp.argotunnel.com).

Again, 1.1.1.1 isn't required, it's simply a fallback in case the regular system resolver isn't working. I'm not sure why your regular system resolver isn't working though!

**Comment by tmsteen** [comment]
Testing on my home network, if I block 1.1.1.1, Argo will not connect. So that may point the actual OS settings I am using. I did not think our provisioning changed any DNS settings (vanilla Ubuntu) but I will investigate that.

**Comment by tanel-andreson** [comment]
This is still open.
If our ISP for some reason is not able to route to 1.1.1.1, then cloudflared service won't start up.
Could you please fall back to 1.0.0.1 as this one seems to not been affected by the overall router bug?

**Comment by TownLake** [comment]
@tanel-andreson thanks for the feedback, we'll review that option on our side. Do you mind sharing the name of your ISP if they are blocking 1.1.1.1, we'd like to reach out to them if that's the case, or are you suggesting a hypothetical where it could occur?

**Comment by tanel-andreson** [comment]
We currently have a problem with AT&T. But when googling I see that this is more related to some specific router models.
We haven't yet got an answer from ATT, but I'll post this when we have more info.

**Comment by tanel-andreson** [comment]
They seem to be ignoring the request to fix their routers to stop using 1.1.1.1
Can you do the workaround of 1.0.0.1 in the code for now?
